pkg/storage/tsdb/block/index.go

```diff
@@ -586,16 +588,16 @@ func rewrite(
 		series   = []seriesRepair{}
 	)
 
-	var lset labels.Labels
+	var builder labels.ScratchBuilder
 	var chks []chunks.Meta
 	for all.Next() {
 		id := all.At()
 
-		if err := indexr.Series(id, &lset, &chks); err != nil {
+		if err := indexr.Series(id, &builder, &chks); err != nil {
```

Comment:
[nit] Same comment here, about adding a comment to explain why we don't need to reset the builder.

Reply:
That seems to be a common pattern for all `IndexReader.Series` calls, I'm not sure adding such comment everywhere makes much sense. It would be better to add such comment to [`IndexReader.Series` method](https://github.com/grafana/mimir/blob/45c47d550b02b1b44e8885596d8db30e0072103b/vendor/github.com/prometheus/prometheus/tsdb/block.go#L93-L96) instead.

Reply:
Fair point Peter. I'm good not adding a comment there.

Reply:
How about https://github.com/grafana/mimir/pull/3912/files#r1066770094 ? (my proposed change above)

Reply:
I think it would make more sense to document in upstream Prometheus's `IndexReader` type, as many implementations come from there. Personally I wouldn't worry about commeint in Mimir codebase.

Reply:
Hmm, this is super confusing, I see implementations for IndexReadertype not doing a reset
e.g. in vendor/github.com/prometheus/prometheus/tsdb/ooo_head_read.go
So to avoid confusion due to the two interfaces (tsdb.IndexReader and indexReader in mimir) describing almost the same thing, maybe we should not be doing the reset inside the function after all. 
@bboreham  ?

Reply:
Isn't it the same as we didn't have to reset `lset` before?

Reply:
I've now checked prometheus and that only has the IndexReader interface and has different behavior in implementations:
[Read.Series](https://github.com/prometheus/prometheus/blob/6948fb1eb68bc0359bb6a3fb74c85e83f56a5c2a/tsdb/index/index.go#L1600) calls [Decoder.Series](https://github.com/prometheus/prometheus/blob/6948fb1eb68bc0359bb6a3fb74c85e83f56a5c2a/tsdb/index/index.go#L1841)  which starts with reset.

But [headIndexRead.Series](https://github.com/prometheus/prometheus/blob/6948fb1eb68bc0359bb6a3fb74c85e83f56a5c2a/tsdb/head_read.go#L153) does not, and neither does OOO Head readers and the mock reader in tests.

So to keep us in sync, should be fixed in prometheus first....